week08/manager.go

```diff
@@ -0,0 +1,231 @@
+
+		e := Employee(i + 1)
+
+		// start the employee working
+		// with the given context and manager
+		go e.work(ctx, m)
+	}
+
+	// return the context for clients to listen to
+	// for cancellation.
+	return ctx, nil
+}
+
+// Assign will assign the given products to employees
+// as employeess become available. An invalid product
+// will return an error.
+func (m *Manager) Assign(products ...*Product) error {
+	if m.stopped {
+		return ErrManagerStopped{}
+	}
+
+	// loop through each product and assign it to an employee
+	for _, p := range products {
+		// validate product
+		if err := p.IsValid(); err != nil {
+			m.Errors() <- err
+			return err
+		}
+
+		// assign product to employee
+		// this will block until an employee becomes available
+		m.Jobs() <- p
+	}
+
+	return nil
+}
+
+// Complete will wrap the employee and the product into
+// a CompletedProduct. The will be passed down the Completed()
+// channel as soon as a listener is available to receive it.
+// Complete will error if the employee is invalid or
+// if the product is not built.
+func (m *Manager) Complete(e Employee, p *Product) error {
+	// validate employee
+	if err := e.IsValid(); err != nil {
+		return err
+	}
+
+	// validate product is built
+	if err := p.IsBuilt(); err != nil {
+		return err
+	}
+
+	cp := CompletedProduct{
+		Employee: e,
+		Product:  *p, // deference pointer to value type ype t
+	}
+
+	// fmt.Printf("TODO >> manager.go:102 cp %[1]T %[1]v\n", cp)
+	// Send completed product to Completed() channel
+	// for a listener to receive it.
+	// This will block until a listener is available.
+	m.completedCh() <- cp
+
+	return nil
+}
+
+// completedCh returns the channel for CompletedProducts
+func (m *Manager) completedCh() chan CompletedProduct {
+	m.Lock()
+	defer m.Unlock()
+
+	if m.completed == nil {
+		m.completed = make(chan CompletedProduct)
+	}
+	return m.completed
+}
+
+// Completed will return a channel that can be listened to
+// for CompletedProducts.
+// This is a read-only channel.
+func (m *Manager) Completed() <-chan CompletedProduct {
+	return m.completedCh()
+}
+
+// Jobs will return a channel that can be listened to
+// for new products to be built.
+func (m *Manager) Jobs() chan *Product {
+	m.Lock()
+	defer m.Unlock()
+
+	if m.jobs == nil {
+		m.jobs = make(chan *Product)
+	}
+
+	return m.jobs
+}
+
+// Errors will return a channel that can be listened to
+// and can be used to receive errors from employees.
+func (m *Manager) Errors() chan error {
+	m.Lock()
+	defer m.Unlock()
+
+	if m.errs == nil {
+		m.errs = make(chan error)
+	}
+	return m.errs
+}
+
+func (m *Manager) IsJobsClosed() bool {
+	select {
+	case <-m.jobs:
+		return true
+	default:
+	}
+
+	return false
+}
+
+func (m *Manager) IsErrorsClosed() bool {
+	select {
+	case <-m.errs:
+		return true
+	default:
+	}
+
+	return false
+}
+
+func (m *Manager) IsCompletedClosed() bool {
+	select {
+	case <-m.completed:
+		return true
+	default:
+	}
+
+	return false
+}
```

Comment:
While, I understand why you wrote these functions, they don't behave the way you think behave, and their use has dangerous side-effects.

The way these methods are coded, if, for example, `IsCompletedClosed` is called and a message is received through the `m.completed` channel that message will be lost. Any time you read a message from a channel, no one else can receive it. These methods, as a result, all have the potential to lose information, which is not good.

Please remove all these methods.